### PR TITLE
Added featureflag

### LINKED ISSUE #8969: Allow test extending OpenSearchSingleNodeTestCase to configure featureFlagSettings similar to OpenSearchIntegTestCase
**Describe the solution you'd like**
We have featureflags mechanism in the core to control enabling/disabling experimental features and configurations related to it. The tests extending from `OpenSearchIntegTestCase` can [configure these featureFlags](https://github.com/opensearch-project/OpenSearch/blob/main/test/framework/src/main/java/org/opensearch/test/OpenSearchIntegTestCase.java#L784) whereas it is not available for tests extending `OpenSearchSingleNodeTestCase`. We should add it over there as well.

### PR DESCRIPTION
**Pull request by stefano-maglione**
Signed-off-by: Stefano Maglione <maglionestefanoinfo@gmail.com>

### Description
Added the featureFlagSettings method to OpenSearchSingleNodeTestCase

### Related Issues
Resolves #8969

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by stefano-maglione** [comment]
[Test Result](https://build.ci.opensearch.org/job/gradle-check/21353/testReport/) (1 failure / +1)
[org.opensearch.client.PitIT.testDeleteAllAndListAllPits](https://build.ci.opensearch.org/job/gradle-check/21353/testReport/junit/org.opensearch.client/PitIT/testDeleteAllAndListAllPits/)

**Comment by stefano-maglione** [comment]
https://github.com/opensearch-project/OpenSearch/issues/8279

File: test/framework/src/main/java/org/opensearch/test/OpenSearchSingleNodeTestCase.java [hunk 1]
```diff
@@ -414,4 +416,19 @@ protected boolean forbidPrivateIndexSettings() {
         return true;
     }
 
+    /**
+     * Setting all feature flag settings at base IT, which can be overridden later by individual
+     * IT classes.
+     *
+     * @return Feature flag settings.
+     */
+    protected Settings featureFlagSettings() {
```
[the remark this answers is not in the document]
Comment (stefano-maglione) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchSingleNodeTestCase.java:426]: Hi, added :)

[the remark this answers is not in the document]
Comment (stefano-maglione) [reply test/framework/src/main/java/org/opensearch/test/OpenSearchSingleNodeTestCase.java:426]: I see some tests failing, not sure they are related

**Comment by stefano-maglione** [comment]
[Test Result](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/) ("6 fallimenti / +5")
[org.opensearch.repositories.azure.AzureBlobStoreRepositoryTests.testWriteRead](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.azure/AzureBlobStoreRepositoryTests/testWriteRead/)
[org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=cluster.allocation_explain/10_basic/cluster shard allocation explanation test with empty request}](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.backwards/MixedClusterClientYamlTestSuiteIT/test__p0_cluster_allocation_explain_10_basic_cluster_shard_allocation_explanation_test_with_empty_request_/)
[org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testRetrieveShallowCopySnapshotCase1](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testRetrieveShallowCopySnapshotCase1/)
[org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testRetrieveShallowCopySnapshotCase2](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testRetrieveShallowCopySnapshotCase2/)
[org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testGetRemoteStoreShallowCopyShardMetadata](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testGetRemoteStoreShallowCopyShardMetadata/)
[org.opensearch.search.SearchServiceTests.testConcurrentSegmentSearchSearchContext](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.search/SearchServiceTests/testConcurrentSegmentSearchSearchContext/)

**Comment by reta** [comment]
> [org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testRetrieveShallowCopySnapshotCase1](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testRetrieveShallowCopySnapshotCase1/)
> [org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testRetrieveShallowCopySnapshotCase2](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testRetrieveShallowCopySnapshotCase2/)
> [org.opensearch.repositories.blobstore.BlobStoreRepositoryTests.testGetRemoteStoreShallowCopyShardMetadata](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.repositories.blobstore/BlobStoreRepositoryTests/testGetRemoteStoreShallowCopyShardMetadata/)
> [org.opensearch.search.SearchServiceTests.testConcurrentSegmentSearchSearchContext](https://build.ci.opensearch.org/job/gradle-check/21474/testReport/junit/org.opensearch.search/SearchServiceTests/testConcurrentSegmentSearchSearchContext/)

Some tests need a small adjustment since those add FeatureFlags using `nodeSettings` (fe `BlobStoreRepositoryTests`):

```
    @Override
    protected Settings nodeSettings() {
        return Settings.builder().put(super.nodeSettings()).put(FeatureFlags.REMOTE_STORE, "true").build();
    }
```

Since we separated them, the `nodeSettings()` override should be replaced with `featureFlagSettings`). Do you need help with thank? Thank you.

**Comment by stefano-maglione** [comment]
@reta added a modification, hope it is fine :)

**Comment by reta** [comment]
> @reta added a modification, hope it is fine :)

@stefano-maglione  could you please rebase against main? (there was a fix for tests merged in), thank you

**Comment by reta** [comment]
Uh ... something went really wrong with rebase @stefano-maglione ...

**Comment by stefano-maglione** [comment]
@reta what went wrong?

**Comment by stefano-maglione** [comment]
step that I did: rebase my forked main with the upstream main (original repo ). Rebase my featureflag branch with my forked main

**Comment by reta** [comment]
> step that I did: rebase my forked main with the upstream main (original repo ). 

I think that should be just simple merge (sync with upstream)

File: server/src/test/java/org/opensearch/search/SearchServiceTests.java [hunk 2]
```diff
@@ -226,7 +226,7 @@ public void onQueryPhase(SearchContext context, long tookInNanos) {
     }
 
     @Override
-    protected Settings nodeSettings() {
+    protected Settings featureFlagSettings() {
```
Comment (reta) [inline server/src/test/java/org/opensearch/search/SearchServiceTests.java:229]: This class needs both, feature flags and settings:

```suggestion
    @Override
    protected Settings featureFlagSettings() {
        return Settings.builder().put(FeatureFlags.CONCURRENT_SEGMENT_SEARCH, true).build();
    }
    
    @Override
    protected Settings nodeSettings() {
         return Settings.builder().put("search.default_search_timeout", "5s").build();
    }
```

File: server/src/test/java/org/opensearch/repositories/blobstore/BlobStoreRepositoryTests.java [hunk 3]
```diff
@@ -106,8 +106,8 @@ protected void assertSnapshotOrGenericThread() {
     }
 
     @Override
-    protected Settings nodeSettings() {
-        return Settings.builder().put(super.nodeSettings()).put(FeatureFlags.REMOTE_STORE, "true").build();
+    protected Settings featureFlagSettings() {
```
Comment (reta) [inline server/src/test/java/org/opensearch/repositories/blobstore/BlobStoreRepositoryTests.java:109]: 👍

**Comment by reta** [comment]
> DCO check is failing sadly, you may need to `git commit --signoff --amend` for the last commit

**Comment by reta** [comment]
Hate to ask @stefano-maglione but the `main` is very fast moving target this days, could you please rebase against `main` again? Please don't forget to add `--signoff` argument to rebase - it should also fix DCO check. Thank you.

File: server/src/test/java/org/opensearch/search/SearchServiceTests.java [hunk 4]
```diff
@@ -225,9 +225,14 @@ public void onQueryPhase(SearchContext context, long tookInNanos) {
         }
     }
 
+    @Override
+    protected Settings featureFlagSettings() {
+        return Settings.builder().put(FeatureFlags.CONCURRENT_SEGMENT_SEARCH, true).build();
```
Comment (sohami) [inline server/src/test/java/org/opensearch/search/SearchServiceTests.java:230]: this should also follow the same convention as other tests (missing the initialization with `super.featureFlagSettings()`)

` Settings.builder().put(super.featureFlagSettings()).put(FeatureFlags.CONCURRENT_SEGMENT_SEARCH, "true").build();`

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@sohami LGTM? thanks!

**Review by sohami — APPROVED** [review]

**Comment by sohami** [comment]
> @sohami LGTM? thanks!

yes thanks. @stefano-maglione DCO check is still failing. You can squash all your commits into one and then add SignOff message in final commit using something like `git commit --amend -s`

**Comment by reta** [comment]
Thanks a lot @stefano-maglione !

**Comment by stefano-maglione** [comment]
Thanks a lot to  you too @reta

**Comment by reta** [comment]
> Thanks a lot to you too @reta . I guess it is all done right?

Yep, all set!

**Pull request merged**